OpenTelemetry (S3M) Databases apps (#1794)
SumoLogic/sumologic-documentation: docs/integrations/databases/opentelemetry/elasticsearch-opentelemetry.md

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Schematics.png')} alt="Schematics" />

## Fields Create in Sumo Logic for Elasticsearch

Following are the [Fields](https://help.sumologic.com/docs/manage/fields/) which will be created as part of Elasticsearch App install if not already present:

- **db.cluster.name** - User configured. Enter a name to identify this Elasticsearch cluster. This cluster name will be shown in the Sumo Logic dashboards.
- **db.system** - Has a fixed value of **elasticsearch**.
- **sumo.datasource** - Has a fixed value of **elasticsearch**.
- **db.node.name** - Has the value of host name of the machine which is being monitored.

### Prerequisites

This receiver supports Elasticsearch versions 7.9+

If Elasticsearch security features are enabled, you must have either the monitor or manage cluster privilege. See the [Elasticsearch docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/authorization.html) for more information on authorization and [Security privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-privileges.html).

> Configure logging in Elasticsearch. Elasticsearch supports logging via local text log files. Elasticsearch logs have four levels of verbosity. To select a level, set loglevel to one of:

Reviewer: ```suggestion
**Configure logging in Elasticsearch.** Elasticsearch supports logging via local text log files. Elasticsearch logs have four levels of verbosity. To select a level, set loglevel to one of:
```

1.  debug: a lot of information, useful for development/testing
2.  verbose: includes information not often needed, but logs less than debug
3.  notice (default value): moderately verbose, ideal for production environments
4.  warning: only very important/critical messages are logged

> All logging settings are located in [Elasticsearch.conf](https://www.elastic.co/guide/en/elasticsearch/reference/current/logging.html). By default, Elasticsearch logs are stored in `/var/log/elasticsearch/ELK-<Clustername>.log`. The default directory for log files is listed in the Elasticsearch.conf file. Logs from the

Reviewer: ```suggestion
All logging settings are located in [Elasticsearch.conf](https://www.elastic.co/guide/en/elasticsearch/reference/current/logging.html). By default, Elasticsearch logs are stored in `/var/log/elasticsearch/ELK-<Clustername>.log`. The default directory for log files is listed in the Elasticsearch.conf file.
```

## Collecting Logs, Metrics & installing Elasticsearch app

Here are the steps for Collecting Logs, metric and installing the app:

### Step 1: Set up Collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the **Use an existing Collector** option.
:::

> If you want to create a new Collector please select **Add a new Collector** option.

Reviewer: ```suggestion
To create a new Collector:
	
	1. Select the **Add a new Collector** option.
	2. Select the platform for which you want to install the Sumo OpenTelemetry Collector.
	
This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.
```

> Select the platform for which you want to install the Sumo OpenTelemetry Collector.

Reviewer: ```suggestion
```

> This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.  

Reviewer: ```suggestion
```

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Collector.png')} alt="Collector" />

### Step 2: Configure integration

In this step we will be configuring the yaml required for Elasticsearch Collection.

Below are the input required:

-   **Endpoint** - Enter the url of the server which needs to be monitored. Example: http://localhost:9200
-   **User Name** - Enter the Elasticsearch Username.
-   **Elasticsearch cluster log path** : By default, Elasticsearch logs are stored in `/var/log/elasticsearch/ELK-<Clustername>.log`.
-   **Tags** : `db.cluster.name`.

> You can include additional metadata which you want to be tagged along with the data ingested to Sumo Logic.

Reviewer: ```suggestion
You can add any custom fields which you want to tag along with the data ingested in sumo.
```

<Tabs
  className="unique-tabs"
  defaultValue="Linux"
  values={[
    {label: 'Linux', value: 'Linux'},
    {label: 'Windows', value: 'Windows'},
    {label: 'macOS', value: 'macOS'},
  ]}>

<TabItem value="Linux">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Elasticsearch instance which needs to be monitored.
2.  Place Env file in the following directory
```sh
/etc/otelcol-sumo/env/
```
3.  restart the collector using
```sh
 sudo systemctl restart otelcol-sumo
```

</TabItem>
<TabItem value="Windows">

1.  Copy the yaml file to `C:\ProgramData\Sumo Logic\OpenTelemetry Collector\config\conf.d` folder in the machine which needs to be monitored.
2.  Restart the collector using 
```sh
Restart-Service -Name OtelcolSumo
```

</TabItem>
<TabItem value="macOS">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Elasticsearch instance which needs to be monitored.
2.  Restart the otelcol-sumo process using the below command 
```sh
 otelcol-sumo --config /etc/otelcol-sumo/sumologic.yaml --config "glob:/etc/otelcol-sumo/conf.d/*.yaml"
```

</TabItem>
</Tabs>

> After successful execution of the above command, Sumo will start receiving the data from your host machine. This will install the app to your Sumo Logic Org. 

Reviewer: ```suggestion
After successful execution of the above command, Sumo will start receiving the data from your host machine. 
Click **Next**. This will install the app to your Sumo Logic Org. 
```

Panels will start to fill automatically. It's important to note that each panel slowly fills with data matching the time range query and received since the panel was created. Results won't immediately be available, but within 20 minutes, you'll see full graphs and maps.

## Sample Log Messages

```sql
{
   "type":"server",
   "timestamp":"2021-07-12T11:42:25,862+07:00",
   "level":"INFO",
   "component":"o.e.x.s.a.s.FileRolesStore",
   "cluster.name":"elasticsearch",
   "node.name":"v103-157-218-134.3stech.vn",
   "message":"parsed [0] roles from file [/etc/elasticsearch/roles.yml]"
}
```

## Sample Open Telemetry metric 

```sql

> {"queryId":"A","_source":"sumo_hosted_collector_otel_elasticsearch","state":"completed","thread_pool_name":"analyze","elasticsearch.node.name":"ip-172-31-86-95","elasticsearch.cluster.name":"elasticsearch","metric":"elasticsearch.node.thread_pool.tasks.finished","db.cluster.name":"elastic_otel_cluster","_collectorId":"000000000C5B7100","deployment.environment":"otel_elastic_dev","_sourceId":"0000000000000000","unit":"{tasks}","db.system":"elasticsearch","_sourceHost":"sumoOtelelasticsearch","_collector":"sumo_hosted_collector_otel_elasticsearch","max":0,"min":0,"avg":0,"sum":0,"latest":0,"count":2}

Reviewer: ```suggestion
{"queryId":"A","_source":"sumo_hosted_collector_otel_elasticsearch","state":"completed","thread_pool_name":"analyze","db.node.name":"ip-172-31-86-95","elasticsearch.cluster.name":"elasticsearch","metric":"elasticsearch.node.thread_pool.tasks.finished","db.cluster.name":"elastic_otel_cluster","_collectorId":"000000000C5B7100","deployment.environment":"otel_elastic_dev","_sourceId":"0000000000000000","unit":"{tasks}","db.system":"elasticsearch","_sourceHost":"sumoOtelelasticsearch","_collector":"sumo_hosted_collector_otel_elasticsearch","max":0,"min":0,"avg":0,"sum":0,"latest":0,"count":2}
```

```

## Sample Log Query

```sql

> Log query from  the panel : Errors

Reviewer: This line should be out of code block

 db.system=elasticsearch %"deployment.environment"={{deployment.environment}} db.cluster.name={{db.cluster.name}} ERROR | json "log" as _rawlog nodrop 
| if (isEmpty(_rawlog), _raw, _rawlog) as _raw
| json field=_raw "timestamp" as timestamp
| json field=_raw "level" as level
| json field=_raw "component" as es_component
| json field=_raw "message" as message
| where level = "ERROR"
| count
```

## Sample Metric Query

Metric query from the panel : JVM Memory Used (MB)
```sql

>  deployment.environment=* metric=jvm.memory.heap.used  db.cluster.name=* elasticsearch.node.name=* | sum by db.cluster.name, elasticsearch.node.name

Reviewer: ```suggestion
 deployment.environment=* metric=jvm.memory.heap.used  db.cluster.name=* db.node.name=* | sum by db.cluster.name, db.node.name
```

If Elasticsearch security features are enabled, you must have either the monitor or manage cluster privilege. See the [Elasticsearch docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/authorization.html) for more information on authorization and [Security privileges](https://www.elastic.co/guide/en/elasticsearch/reference/current/security-privileges.html).

**Configure logging in Elasticsearch.** Elasticsearch supports logging via local text log files. Elasticsearch logs have four levels of verbosity. To select a level, set loglevel to one of:

1.  debug: a lot of information, useful for development/testing
2.  verbose: includes information not often needed, but logs less than debug
3.  notice (default value): moderately verbose, ideal for production environments
4.  warning: only very important/critical messages are logged

All logging settings are located in [Elasticsearch.conf](https://www.elastic.co/guide/en/elasticsearch/reference/current/logging.html). By default, Elasticsearch logs are stored in `/var/log/elasticsearch/ELK-<Clustername>.log`. The default directory for log files is listed in the Elasticsearch.conf file.

> ## Collecting Logs, Metrics & installing Elasticsearch app

Reviewer 2: ```suggestion
## Collecting Logs, Metrics and installing Elasticsearch app
```

Here are the steps for Collecting Logs, metric and installing the app:

### Step 1: Set up Collector

:::note
If you want to use an existing OpenTelemetry Collector, you can skip this step by selecting the **Use an existing Collector** option.
:::

To create a new Collector:
	
	1. Select the **Add a new Collector** option.
	2. Select the platform for which you want to install the Sumo OpenTelemetry Collector.
	
This will generate a command which can be executed in the machine which needs to get monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Collector.png')} alt="Collector" />

Reviewer 2: This image link is broken.

### Step 2: Configure integration

In this step we will be configuring the yaml required for Elasticsearch Collection.

Below are the input required:

-   **Endpoint** - Enter the url of the server which needs to be monitored. Example: http://localhost:9200
-   **User Name** - Enter the Elasticsearch Username.
-   **Elasticsearch cluster log path** : By default, Elasticsearch logs are stored in `/var/log/elasticsearch/ELK-<Clustername>.log`.
-   **Tags** : `db.cluster.name`.

> You can add any custom fields which you want to tag along with the data ingested in sumo.

Reviewer 2: ```suggestion
You can add any custom fields which you want to tag along with the data ingested in Sumo.
```

Click on the **Download YAML File** button to get the yaml file.

For linux platform - Click on **Download Environment Variables File** button to get the file with the password which is supposed to be set as environment variable.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-YAML.png')} alt="YAML" />

Reviewer 2: This image link is broken.

```

## Sample Log Query

```sql
Log query from  the panel : Errors
 db.system=elasticsearch %"deployment.environment"={{deployment.environment}} db.cluster.name={{db.cluster.name}} ERROR | json "log" as _rawlog nodrop 
| if (isEmpty(_rawlog), _raw, _rawlog) as _raw
| json field=_raw "timestamp" as timestamp
| json field=_raw "level" as level
| json field=_raw "component" as es_component
| json field=_raw "message" as message
| where level = "ERROR"
| count
```

## Sample Metric Query

Metric query from the panel : JVM Memory Used (MB)
```sql
 deployment.environment=* metric=jvm.memory.heap.used  db.cluster.name=* db.node.name=* | sum by db.cluster.name, db.node.name
```

## Viewing Elasticsearch Dashboards

### Overview

The **Elasticsearch - Overview** dashboard provides the health of Elasticsearch clusters, shards analysis, resource utilization of Elasticsearch host & clusters, search and indexing performance.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Overview.png')} alt="Overview" />

Reviewer 2: This image link is broken.

### Total Operations Stats

The **Elasticsearch - Total Operations stats** dashboard provides information on the operations of the Elasticsearch system.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Total-Operations-Stats.png')} alt="Total Operations Stats" />

Reviewer 2: This image link is broken.

### Thread Pool

The **Elasticsearch - Thread Pool** dashboard analyzes thread pools operations to manage memory consumption of nodes in the cluster.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Thread-Pool.png')} alt="Thread Pool" />

### Resource

The **Elasticsearch - Resource** dashboard monitors JVM Memory, Network, Disk, Network and CPU of Elasticsearch node.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Resource.png')} alt="Resource" />

Reviewer 2: This image link is broken.

### Performance Stats

The **Elasticsearch - Performance Stats** dashboard performance statistics such as latency and Translog operations and size.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Performance-Stats.png')} alt="Performance Stats" />

Reviewer 2: This image link is broken.

### Indices

The **Elasticsearch - Indices** dashboard monitors Index operations, size and latency. It also provides analytics on doc values, fields, fixed bitsets, and terms memory.

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Indices.png')} alt="Indices" />

### Documents

The **Elasticsearch - Documents** dashboard provides analytics and monitoring on Elasticsearch documents.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Documents.png')} alt="Documents" />

Reviewer 2: This image link is broken.

### Caches

The **Elasticsearch - Caches** dashboard allows you to monitor query cache size, evictions and field data memory size.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Caches.png')} alt="Caches" />

Reviewer 2: This image link is broken.

### Errors And Warnings

The **ElasticSearch - Errors And Warnings** dashboard shows errors and warnings by Elasticsearch components.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Errors-And-Warnings.png')} alt="Errors And Warnings" />

Reviewer 2: This image link is broken.

### Garbage Collection

The **Elasticsearch - Garbage Collector** dashboard provides information on the garbage collection of the Java Virtual Machine.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Garbage-Collection.png')} alt="Garbage Collection" />

Reviewer 2: This image link is broken.

### Login And Connections

The **ElasticSearch - Login And Connections** dashboard shows geo location of client connection requests, failed connection logins and count of failed login attempts

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/ElasticSearch-OpenTelemetry/ElasticSearch-Login-And-Connections.png')} alt="Login And Connections" />

Reviewer 2: This image link is broken.
